### PR TITLE
feat(botocore): patch stepfunctions

### PR DESCRIPTION
**Pull request by agocs**
Patch botocore stepfunctions start-execution and start-sync-execution. We want to be able to add trace context to the `input` so the stepfunction logs trace reducer can link stepfunction traces to upstream traces.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

**Comment by majorgreys** [comment]
@agocs Can you add a release note for this feature change?

https://ddtrace.readthedocs.io/en/stable/releasenotes.html#generating-a-release-note

File: releasenotes/notes/inject-botocore-stepfunction-start_execution-calls-95bed0ca2e1d006e.yaml
Comment (majorgreys) [inline releasenotes/notes/inject-botocore-stepfunction-start_execution-calls-95bed0ca2e1d006e.yaml:4]:
```suggestion
    botocore: Add the ability to inject trace context into the input field of botocore stepfunction start_execution and 
```

**Review by ZStriker19 — APPROVED** [review]
Is extraction already implemented?

**Comment by ZStriker19** [comment]
> Is extraction already implemented?

Oh or there is no extraction?  Because "stepfunction logs trace reducer can link stepfunction traces to upstream traces."

**Comment by agocs** [comment]
@ZStriker19 That's correct, extraction will be implemented in the logs-to-traces reducer

**Review by purple4reina — CHANGES_REQUESTED** [review]
Let's do a ducktype check before adding the dd header dict.

Also, looks like we could use more test coverage of this `inject_trace_to_stepfunction_input`. I'm seeing 5 different code paths, yet only one test.

File: ddtrace/contrib/botocore/patch.py
Comment (agocs) [reply ddtrace/contrib/botocore/patch.py]: Good catch! Thanks. Input validation for start_execution is weird and inconsistent, so I'm trying to cast as broad a net as possible. I added additional tests for different possible data types.

File: tests/contrib/botocore/test.py
Comment (agocs) [reply tests/contrib/botocore/test.py:941]: You're not wrong!

**Review by purple4reina — APPROVED** [review]

**Review by emmettbutler — APPROVED** [review]

**Comment by agocs** [comment]
I think I'm going to wait until after the code freeze to merge this rather than rush to get it in before EOY.

**Pull request merged**
